Feature changes

Incompleteusern opened the pull request:
This pr rewrites `Feature` to be more general by removing the `apis` and `keybinds` list, instead letting subclasses load them if necessary. Furthermore, it shifts `WebManager` to statically loads apis as well.

This is currently a draft due to depending on #13 as well as being as of now untested

Incompleteusern committed:
Many fixes

Incompleteusern committed:
statically load apis and generify Feature

Incompleteusern committed:
spotless

Incompleteusern committed:
reorganize method

magicus commented on common/src/main/java/com/wynntils/core/features/Feature.java:
```diff
@@ -48,10 +39,15 @@ public Feature() {
-            ImmutableList.Builder<WebProviderSupplier> apis,
-            ImmutableList.Builder<KeySupplier> keybinds,
-            ImmutableList.Builder<Condition> conditions) {
+    protected void init(ImmutableList.Builder<Condition> conditions) {
```
You can make a method abstract to force subclasses to implement it...

Incompleteusern replied:
I want to make it optional to extend

magicus commented on common/src/main/java/com/wynntils/core/features/Feature.java:
```diff
@@ -60,48 +56,14 @@ protected void init(
      */
-    public final boolean enable() {
+    public final void enable() {
         if (enabled)
```
I realize this is not new code, but I really really dislike the use of multiline if statements without braces. I'd like to vote for having a formal rule against it; preferably checked by style checkers if possible...

Incompleteusern replied:
I’m fine either way, get a consensus and I’ll follow the guideline

magicus commented on common/src/main/java/com/wynntils/core/webapi/WebManager.java:
```diff
@@ -28,45 +26,37 @@
+                        .cacheTo(new File(API_CACHE_ROOT, "item_guesses.json"))
+                        .handleJsonObject(
+                                json -> {
+                                    Type type =
```
spotless is really having a weird opinion on how good code should look. Maybe we should go back to the drawing board and start looking for another formatter. This looks horrible and almost unreadable, imho.

Incompleteusern replied:
Yeah, lambdas/builders aren’t handled well

magicus reviewed (commented):
Overall it seems like a reasonable refactoring. Bit hard to tell due to #13 mixed in with this.

Incompleteusern committed:
Merge branch 'main' into feature

Incompleteusern commented:
Yeah the only real changes are in [082da0](https://github.com/Wynntils/Artemis/pull/14/commits/082da0df7df12c3128adc6748b38f911fbdd9783)

Incompleteusern committed:
Changes

- Revert back to on demand feature loading
- Force extension of feature methods
- Don't automatically register class
- Clean up static registers

Incompleteusern committed:
spotless

Incompleteusern commented:
Nvm, the new commit of [7e0166](https://github.com/Wynntils/Artemis/pull/14/commits/7e0166fb53b7446204db17bb3231bc9d25943fa0) also changes stuff

kristofbolyai reviewed (approved):
Looks good to me!

magicus reviewed (approved):
I'm getting a bit lost in all changes, but I think this looks good. Anyway, push it, and if it needs to be polished, we can do that later.

In general, I think this is a sane way of developing -- as we notice that more structured APIs are needed, we refactor the code base to get there. It's hard to get a proper architecture correct in advance of actually using it.
